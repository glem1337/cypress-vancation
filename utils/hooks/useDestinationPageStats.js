import { useSelector, useDispatch } from 'react-redux';

import { MAPBOX_FEATURE_TYPE } from 'constants/searchDestinations';
import {
  searchDestinationParamsSelector,
  searchDestinationFiltersSelector,
  desktopMapVisibilitySelector,
} from 'state/concepts/search-destinations/selectors';
import isPresent from 'utils/isPresent';
import {
  clearAllFilter as clearAllFilterAction,
  toggleDesktopMapVisibility as toggleDesktopMapVisibilityAction,
} from 'state/concepts/search-destinations/actions';
import { loadingSelector } from 'state/data/selectors';
import { fetchCampersEndpoint } from 'state/concepts/campervan-rental/endpoints';
import { campersSelector } from 'state/concepts/campervan-rental/selectors';

function useDestinationPageStats() {
  const dispatch = useDispatch();

  const campers = useSelector(campersSelector);

  const searchParams = useSelector(searchDestinationParamsSelector);

  const persistedFilters = useSelector(searchDestinationFiltersSelector);

  const isDesktopMapVisible = useSelector(desktopMapVisibilitySelector);

  const areCampersFetching = useSelector(
    state => loadingSelector(state, fetchCampersEndpoint.endpoint),
  );

  /**
   * Is any filter applied.
   */
  const isAnyFilterApplied = () => {
    const keys = Object.keys(persistedFilters);

    for (let i = 0; i < keys.length; i += 1) {
      const key = keys[i];

      if (
        key === 'uuid'
        || key === 'isMobileVisible'
      ) {
        // eslint-disable-next-line no-continue
        continue;
      }

      const value = persistedFilters[key];

      if (isPresent(value)) {
        return true;
      }
    }

    return false;
  };

  /**
   * Clear all filters.
   */
  const clearAllFilters = () => {
    dispatch(clearAllFilterAction());
  };

  /**
   * Toggle desktop map visibility
   */
  const toggleDesktopMapVisibility = () => {
    dispatch(toggleDesktopMapVisibilityAction());
  };

  return {
    isSearchResultsPage: searchParams?.location?.type === MAPBOX_FEATURE_TYPE,
    isAnyFilterApplied: isAnyFilterApplied(),
    clearAllFilters,
    isDesktopMapVisible,
    toggleDesktopMapVisibility,
    areCampersFetching,
    campers,
  };
}

export default useDestinationPageStats;
