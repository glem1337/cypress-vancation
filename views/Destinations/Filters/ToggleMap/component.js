import React from 'react';
import { FormattedMessage } from 'react-intl';
import { Switch } from 'antd';

import useDestinationPageStats from 'utils/hooks/useDestinationPageStats';

const ToggleMap = () => {
  const {
    isSearchResultsPage,
    isDesktopMapVisible,
    toggleDesktopMapVisibility,
  } = useDestinationPageStats();

  if (!isSearchResultsPage) {
    return null;
  }

  return (
    <div className="d-flex align-items-center ml-auto">
      <span className="mr-8">
        <FormattedMessage id="shared.mapView" />
      </span>
      <Switch
        defaultChecked={isDesktopMapVisible}
        onChange={toggleDesktopMapVisibility}
      />
    </div>
  );
};

export default ToggleMap;
