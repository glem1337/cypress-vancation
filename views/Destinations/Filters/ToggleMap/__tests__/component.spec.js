import React from 'react';
import { shallow } from 'enzyme';

import useDestinationPageStats from 'utils/hooks/useDestinationPageStats';

import ToggleMap from '../component';

jest.mock('utils/hooks/useDestinationPageStats', () => jest.fn(() => ({
  isSearchResultsPage: true,
  isDesktopMapVisible: true,
  toggleDesktopMapVisibility: jest.fn(),
})));

describe('ToggleMap component tests', () => {
  let component = shallow(<ToggleMap />);

  it('matches snapshot', () => {
    expect(component).toMatchSnapshot();
  });

  it('matches snapshot when isSearchResultsPage equals false', () => {
    useDestinationPageStats.mockReturnValueOnce({
      isSearchResultsPage: false,
      isDesktopMapVisible: true,
      toggleDesktopMapVisibility: jest.fn(),
    });

    component = shallow(<ToggleMap />);

    expect(component).toMatchSnapshot();
  });
});
