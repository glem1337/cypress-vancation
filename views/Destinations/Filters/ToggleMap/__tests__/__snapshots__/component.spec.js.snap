// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleMap component tests matches snapshot 1`] = `
<div
  className="d-flex align-items-center ml-auto"
>
  <span
    className="mr-8"
  >
    <Memo(FormattedMessage)
      id="shared.mapView"
    />
  </span>
  <Switch
    defaultChecked={true}
    onChange={[MockFunction]}
  />
</div>
`;

exports[`ToggleMap component tests matches snapshot when isSearchResultsPage equals false 1`] = `""`;
